src/DIRAC/DataManagementSystem/DB/FTS3DB.py

```diff
@@ -537,12 +544,13 @@ def getNonFinishedOperations(self, limit=20, operationAssignmentTag="Assigned"):
             ftsOperations = []
 
             # We need to do the select in two times because the join clause that makes the limit difficult
+            # We get the list of operations ID that have associated jobs assigned
+            opIDsWithJobAssigned = session.query(FTS3Job.operationID).filter(~FTS3Job.assignment.is_(None)).subquery()
             operationIDsQuery = (
                 session.query(FTS3Operation.operationID)
-                .outerjoin(FTS3Job)
                 .filter(FTS3Operation.status.in_(["Active", "Processed"]))
                 .filter(FTS3Operation.assignment.is_(None))
-                .filter(FTS3Job.assignment.is_(None))
+                .filter(~FTS3Operation.operationID.in_(opIDsWithJobAssigned))
```

Comment:
This fixes the race condition? Is the splitting of monitoring into batches still needed to fix the problem?

Reply:
This fixes the race condition.
Splitting into batches optimizes even further because we do not block at once many jobs/operations.